pkg/acquisition/modules/journalctl/journalctl.go

```diff
@@ -0,0 +1,204 @@
+
+import (
+	"bufio"
+	"fmt"
+	"net/url"
+	"os/exec"
+	"strings"
+	"time"
+
+	"github.com/crowdsecurity/crowdsec/pkg/acquisition/configuration"
+	leaky "github.com/crowdsecurity/crowdsec/pkg/leakybucket"
+	"github.com/crowdsecurity/crowdsec/pkg/types"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	log "github.com/sirupsen/logrus"
+
+	"gopkg.in/tomb.v2"
+	"gopkg.in/yaml.v2"
+)
+
+type JournalCtlConfiguration struct {
+	configuration.DataSourceCommonCfg `yaml:",inline"`
+	Filters                           []string `yaml:"journalctl_filter"`
+}
+
+type JournalCtlSource struct {
+	config JournalCtlConfiguration
+	logger *log.Entry
+	src    string
+	args   []string
+}
+
+const journalctlCmd string = "journalctl"
+
+var (
+	journalctlArgsOneShot  = []string{""}
+	journalctlArgstreaming = []string{"--follow"}
+)
+
+func (j *JournalCtlSource) runJournalCtl(out chan types.Event, t *tomb.Tomb) error {
+	cmd := exec.Command(journalctlCmd, j.args...)
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return fmt.Errorf("could not get journalctl stdout: %s", err)
+	}
+	stderr, err := cmd.StderrPipe()
+	if err != nil {
+		return fmt.Errorf("could not get journalctl stderr: %s", err)
+	}
+
+	readErr := make(chan error)
+
+	j.logger.Debugf("Running journalctl command: %s %s", cmd.Path, cmd.Args)
+	err = cmd.Start()
+	if err != nil {
+		j.logger.Errorf("could not start journalctl command : %s", err)
+		return err
+	}
+
+	go func() {
+		scanner := bufio.NewScanner(stderr)
+		if scanner == nil {
+			readErr <- fmt.Errorf("failed to create stderr scanner")
+			return
+		}
+		for scanner.Scan() {
+			txt := scanner.Text()
+			j.logger.Warningf("got stderr message : %s", txt)
+			readErr <- fmt.Errorf(txt)
+		}
+	}()
+
+	go func() {
+		scanner := bufio.NewScanner(stdout)
+		if scanner == nil {
+			readErr <- fmt.Errorf("failed to create stdout scanner")
+			return
+		}
+		for scanner.Scan() {
+			l := types.Line{}
+			//ReaderHits.With(prometheus.Labels{"source": j.SrcName}).Inc()
+			l.Raw = scanner.Text()
+			j.logger.Debugf("getting one line : %s", l.Raw)
+			l.Labels = j.config.Labels
+			l.Time = time.Now()
+			l.Src = j.src
+			l.Process = true
+			l.Module = j.GetName()
+			evt := types.Event{Line: l, Process: true, Type: types.LOG, ExpectMode: leaky.LIVE}
+			out <- evt
+		}
+		j.logger.Debugf("finished reading from journalctl")
+		if err := scanner.Err(); err != nil {
+			j.logger.Debugf("got an error while reading %s : %s", j.src, err)
+			readErr <- err
+			return
+		}
+		readErr <- nil
+	}()
+
+	for {
+		select {
+		case <-t.Dying():
+			j.logger.Debugf("journalctl datasource %s stopping", j.src)
+			return nil
+		case err := <-readErr:
+			j.logger.Debugf("the subroutine returned, leave as well")
+			if err != nil {
+				j.logger.Warningf("journalctl reader error : %s", err)
+				t.Kill(err)
+			}
+			return err
+		}
+	}
+}
+
+func (j *JournalCtlSource) GetMetrics() []prometheus.Collector {
```

Comment:
Is there reeally no way to have some metrics here ?

Reply:
Possible, just not implemented yet.